File: algorithms/python/QuickSort.py
# an implementation of quicksort. the last element is used as the pivot. later we'll randomize the pivot choice.

# Partitions the subarray A[p..r] of A, using the last element A[r] as the pivot. Elements less
#   than A[r] go to the left, and elements greater than or equal to the pivot go to the right.
# Runtime: O(n)
def Partition(A, p, r):
    i = p-1
    for j in range(p,r):
        if A[j] < A[r]:     # last element A[r] is used as the pivot
            i+=1
            A[i],A[j] = A[j],A[i]
    A[i+1],A[r] = A[r],A[i+1]
    return i+1

# Worst case runtime: O(n^2) i.e. when the partition is always 1 to n-1
# Best case runtime: O(n*logn) i.e. when the partition is always a constant ratio like 1/2 to 1/2 or 9/10 to 1/10
# We can get O(n*logn) EXPECTED runtime by randomizing the pivot
def QuickSort(A, p, r):
    if p < r:
        q = Partition(A, p, r)
        QuickSort(A, p, q-1)
        QuickSort(A, q+1, r)

"""
#Let's test out QuickSort before randomizing the pivot.
from sys import argv
argNums = [int(x) for x in argv[1:]]
QuickSort(argNums, 0, len(argNums)-1)
print argNums
"""

# Now let's randomize the pivot
# Runtime: O(n)
from random import random
def RPartition(A, p, r):
    i = p + int((r-p+1)*random()) #pick a random index from [p..r]
    A[i],A[r] = A[r],A[i]    # ..and swap that element with the last one
    return Partition(A, p, r)       # ..then just call Partition!

# The new version of quicksort just uses RPartition instead of Partition.
# Expected runtime: O(n*logn)
def RQuickSort(A, p, r):
    if p < r:
        q = RPartition(A, p, r)
        QuickSort(A, p, q-1)
        QuickSort(A, q+1, r)

# Everything should still work, but let's check again anyway.
from sys import argv
argNums = [int(x) for x in argv[1:]]
#RQuickSort(argNums,0,len(argNums)-1)
RPartition(argNums,0,len(argNums)-1)
print argNums


# Good to go!

File: CSCI382/c++/QuickSort.cpp
/* An implementation of QuickSort. The first version uses the last element of A as the pivot. The second version
 * uses a random pivot to improve expected runtime.
 */
#include <iostream>
#include <vector>
#include <random>
using namespace std;

int Partition(vector<int> &A, int p, int r);

void Swap(vector<int> &A, int i, int j);

// Uses Partition, which uses the last element as the pivot.
// Worst case runtime: O(n^2) when partition does an n-1 to 1 split every time.
// Best case runtime: O(nlogn) when partition does the same constant ratio split every time.
void QuickSort(vector<int> &A, int p, int r)
{
    if (p<r)
    {
        int q = Partition(A, p, r);
        QuickSort(A, p, q-1);
        QuickSort(A, q+1, r);
    }
}

// Uses the last element as the pivot.
// Runtime O(n).
int Partition(vector<int> &A, int p, int r)
{
    int i = p-1;
    for (int j=p; j<r; j++)
    {
        if (A[j] < A[r])
        {
            i++;
            Swap(A, i, j);
        }
    }
    Swap(A, i+1, r);
    return i+1;
}

// Swaps two elements in the vector A.
void Swap(vector<int> &A, int i, int j)
{
    int c = A[i];
    A[i] = A[j];
    A[j] = c;
}

// We can rewrite partition to use a random pivot, which then gives quicksort an expected runtime <O(nlogn)>.
// Runtime: O(n)
int RPartition(vector<int> &A, int p, int r)
{
    // this is best way I've seen so far to generate a random number in [p..r] (includes endpoints!)
    random_device rd; //the seed
    mt19937 gen(rd()); // the generator
    uniform_int_distribution<int> dis(p,r); // the distribution
    int i = dis(gen); // finally, the random index
    Swap(A,i,r);
    return Partition(A,p,r); // we just call the old Partition on the adjusted A.
}

// Now we can rewrite quicksort using RPartition.
// Expected runtime: <O(nlogn)>
void RQuickSort(vector<int> &A, int p, int r)
{
    if (p<r)
    {
        int q = RPartition(A,p,r);
        RQuickSort(A,p,q-1);
        RQuickSort(A, q+1, r);
    }
}

// Now let's test everything out on the command line.
int main(int argc, char **argv)
{
    vector<int> A(argc-1);
    for (int i=0; i<argc-1; i++)
    {
        A[i] = atoi(argv[i+1]);
    }
    RQuickSort(A,0,A.size()-1);
    for (auto x: A) cout << x << " ";
    cout << endl;
    return 0;
}

// Seems to work fine!
